### PR TITLE
[IMP] mail: add hover download button on image attachments

### PR DESCRIPTION
**Pull request by rapr-odoo**
**Current behavior before PR:**

Improve how image attachments are downloaded from the chatter by adding the
`download` button when hovering image files

**Desired behavior after PR is merged:**

add a `Download` button (icon) on the bottom right corner of image attachments
on hover

Task-2802810

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mail/static/tests/qunit_suite_tests/components/message_tests.js [hunk 1]
```diff
@@ -699,6 +699,35 @@ QUnit.test('prevent attachment delete on non-authored message in channels', asyn
...
+        document.body,
+        '.o_AttachmentImage',
+        "should have an attachment",
+    );
+    assert.containsOnce(
+        document.body,
+        '.o_AttachmentImage_actionDownload',
+        "should have download attachment button"
+    );
+});
```
Comment (alexkuhn) [inline addons/mail/static/tests/qunit_suite_tests/components/message_tests.js:729]: We are currently changing tests so that:
- Most data are made `MockServer`-side, with `pyEnv` (instead of locally with `messaging.models[].create()`, because the server has autority on existence of message)
- Make test with a root component like `Discuss` (instead of creating a small component like `createMessageComponent`, because this can easily break due to business code relying on some context of message being viewed).

File: addons/mail/static/src/models/attachment_image.js [hunk 2]
```diff
@@ -8,6 +8,18 @@ registerModel({
...
     identifyingFields: ['attachmentList', 'attachment'],
     recordMethods: {
+        /**
+         * Called when clicking on download icon.
+         *
+         * @param {MouseEvent} ev
+         */
+        onClickDownload(ev) {
+            ev.stopPropagation(); // prevents from opening viewer
+            if (!this.attachment) {
```
Comment (alexkuhn) [inline addons/mail/static/src/models/attachment_image.js:18]: `attachment` is a mandatory field, because this is a required identifying field.
What you need is `!this.exists(0` instead.

File: addons/mail/static/src/models/attachment_image.js [hunk 3]
```diff
@@ -8,6 +8,18 @@ registerModel({
     name: 'AttachmentImage',
     identifyingFields: ['attachmentList', 'attachment'],
     recordMethods: {
+        /**
+         * Called when clicking on download icon.
+         *
+         * @param {MouseEvent} ev
+         */
+        onClickDownload(ev) {
+            ev.stopPropagation(); // prevents from opening viewer
```
Comment (alexkuhn) [inline addons/mail/static/src/models/attachment_image.js:17]: Please don't use `stropPropagation()`, and instead `markEventAsHandled`.
If you want to prevent opening viewer, you should modify handler method of opening viewer to pre-emptively return if it's handled by click download.

The reason we prefer marking event is because it's much easier for traceability of business logic. `stropPropagation` is too strong and is very hard to trace while debugging, because `stropPropagation` kills the event as if it never existed for the following process of dispatching event (i.e. rest of capturing and bubbling phase).

File: addons/mail/static/src/components/attachment_image/attachment_image.xml [hunk 4]
```diff
@@ -20,15 +20,17 @@
...
+                    <div class="o_AttachmentImage_imageOverlay position-absolute top-0 bottom-0 start-0 end-0 p-2 text-white opacity-0 opacity-100-hover d-flex align-items-end  flax-wrap flex-column">
+                        <div t-if="attachmentImage.attachment.isDeletable" class="o_AttachmentImage_action o_AttachmentImage_actionUnlink btn btn-sm btn-dark rounded opacity-75 opacity-100-hover"
                                 t-att-class="{'o-pretty': attachmentImage.attachmentList.composerViewOwner}" t-on-click="attachmentImage.onClickUnlink" title="Remove"
                             >
                                 <i class="fa fa-trash"/>
-                            </div>
                         </div>
-                    </t>
+                        <div t-if="!attachmentImage.attachmentList.composerViewOwner and !attachmentImage.attachment.isUploading" class="o_AttachmentImage_action o_AttachmentImage_actionDownload btn btn-sm btn-dark rounded opacity-75 opacity-100-hover mt-auto" t-on-click="attachmentImage.onClickDownload" title="Download"
```
Comment (alexkuhn) [inline addons/mail/static/src/components/attachment_image/attachment_image.xml:29]: Introduce a new concept on models `AttachmentImage`, by means of a computed field. This field can be named `hasDownloadButton`. That way, here it's just `attachmentImage.hasDownloadButton`, and the condition of presence of this button is abstracted in the computation of the `hasDownloadButton` field.

File: addons/mail/static/src/components/attachment_image/attachment_image.xml [hunk 5]
```diff
@@ -20,15 +20,17 @@
                             <i class="fa fa-spin fa-spinner"/>
                         </div>
                     </t>
-                    <t t-if="attachmentImage.attachment.isDeletable">
-                        <div class="o_AttachmentImage_imageOverlay position-absolute top-0 bottom-0 start-0 end-0 text-right p-2 text-white opacity-0 opacity-100-hover">
-                            <div class="o_AttachmentImage_action o_AttachmentImage_actionUnlink btn btn-sm btn-dark rounded opacity-75 opacity-100-hover"
+                    <div class="o_AttachmentImage_imageOverlay position-absolute top-0 bottom-0 start-0 end-0 p-2 text-white opacity-0 opacity-100-hover d-flex align-items-end  flax-wrap flex-column">
```
Comment (alexkuhn) [inline addons/mail/static/src/components/attachment_image/attachment_image.xml:23]: single whitespace between classnames (there are 2 between `align-items-end` and `flax-wrap`)

**Review by alexkuhn — CHANGES_REQUESTED** [review]

File: addons/mail/static/tests/qunit_suite_tests/components/message_tests.js (hunk 1 above)
Comment (rapr-odoo) [reply addons/mail/static/tests/qunit_suite_tests/components/message_tests.js]: @alexkuhn Changes Done!

File: addons/mail/static/tests/qunit_suite_tests/components/message_tests.js [hunk 6]
```diff
@@ -697,6 +697,46 @@ QUnit.test('prevent attachment delete on non-authored message in channels', asyn
...
+    const thread = messaging.models['Thread'].findFromIdentifyingData({
+        id: mailChannelId1,
+        model: 'mail.channel',
+    });
+    const threadViewer = messaging.models['ThreadViewer'].create({
+        hasThreadView: true,
+        qunitTest: insertAndReplace(),
+        thread: replace(thread),
+    });
+    await createThreadViewComponent(threadViewer.threadView);
```
Comment (alexkuhn) [inline addons/mail/static/tests/qunit_suite_tests/components/message_tests.js:726]: Prefer using a main component, such as `Discuss` with `openDiscuss` helper returned by `start()`.

We are currently working in changing tests so that they no longer mount small components like that.
Reasoning is that these tests are fragile, because it mounts a components without an actual and legit context that makes sense. So it can work by chance / thanks to implementation details, but would easily fail with small unrelated changes.

File: addons/mail/static/src/models/attachment_image.js [hunk 7]
```diff
@@ -99,6 +130,12 @@ registerModel({
             readonly: true,
             required: true,
         }),
+        /**
+         * Determines whether `this` should display a download button.
+         */
+        hasDownloadButton: attr({
+            compute: '_computeHasDownloadButton',
+        }),
```
Comment (alexkuhn) [inline addons/mail/static/src/models/attachment_image.js:138]: add this option: `default: false,`, so that this is `false` by default

File: addons/mail/static/src/models/attachment_image.js [hunk 8]
```diff
@@ -36,6 +57,16 @@ registerModel({
                 this.update({ attachmentDeleteConfirmDialog: insertAndReplace() });
             }
         },
+        /**
+         * @private
+         * @returns {boolean}
+         */
+        _computeHasDownloadButton() {
+            if (!this.attachment || !this.attachmentList) {
+                return false;
```
Comment (alexkuhn) [inline addons/mail/static/src/models/attachment_image.js:66]: `return clear()` (so that it makes use of `default`).

Reasoning for using `default` rather than explicit `false` is because they do not say the same thing: `false` is an intentional and willing decision to not show the download button. Whereas using `default` is more like "I don't care / it doesn't matter / it doesn't apply / etc.".

Choosing a default is to take into account this "non-choice". We want to change it and keep the same intention of "I don't care / it doesn't matter / it doesn't apply / etc.".

File: addons/mail/static/src/components/attachment_image/attachment_image.xml [hunk 9]
```diff
@@ -20,15 +20,17 @@
...
+                    <div class="o_AttachmentImage_imageOverlay position-absolute top-0 bottom-0 start-0 end-0 p-2 text-white opacity-0 opacity-100-hover d-flex align-items-end flax-wrap flex-column">
+                        <div t-if="attachmentImage.attachment.isDeletable" class="o_AttachmentImage_action o_AttachmentImage_actionUnlink btn btn-sm btn-dark rounded opacity-75 opacity-100-hover"
                                 t-att-class="{'o-pretty': attachmentImage.attachmentList.composerViewOwner}" t-on-click="attachmentImage.onClickUnlink" title="Remove"
                             >
                                 <i class="fa fa-trash"/>
-                            </div>
                         </div>
-                    </t>
+                        <div t-if="attachmentImage.hasDownloadButton" class="o_AttachmentImage_action o_AttachmentImage_actionDownload btn btn-sm btn-dark rounded opacity-75 opacity-100-hover mt-auto" t-on-click="attachmentImage.onClickDownload" title="Download"
+                        >
```
Comment (alexkuhn) [inline addons/mail/static/src/components/attachment_image/attachment_image.xml:30]: `>` inline.

Don't change in `actionUnlink` above, just to not make an unnecessary diff.

**Review by alexkuhn — CHANGES_REQUESTED** [review]

File: addons/mail/static/tests/qunit_suite_tests/components/message_tests.js (hunk 6 above)
Comment (rapr-odoo) [reply addons/mail/static/tests/qunit_suite_tests/components/message_tests.js]: @alexkuhn Changes Done!

**Review by alexkuhn — APPROVED** [review]
@robodoo r+

**Comment by seb-odoo** [comment]
@robodoo retry
